### PR TITLE
Indexing tuple intersection type beyond length produces unexpected type

### LINKED ISSUE #42557: Indexing tuple intersection type beyond length produces unexpected type
# Bug Report

### 🔎 Search Terms

tuple intersection index type

### ⏯ Playground Link

[Playground Link](https://www.typescriptlang.org/play?ssl=6&ssc=1&pln=5&pc=1#code/C4TwDgpgBAKlC8UDaA7ArgWwEYQE4BoosB7YgGwgEMUBdKAMigG8oAPALigGdhcBLFAHMoAXwDcAKAAmEAMZlKuaLOIoeUYJxiSVa4FEoINSAAw0xUAPSWo6bHgm71WI8CQBGc1ZslyVFI6q6rKuSABMXta2mDi4UAA+RKQU1FAAhAD8EkA)

### 💻 Code

```ts
type T = [number, boolean] & { x: string };
declare const t: T;
const a = t[0]; // number
const b = t[1]; // boolean
const c = t[2]; // number | boolean !?
```

### 🙁 Actual behavior

`c` is typed as `number | boolean`

### 🙂 Expected behavior

`c` should be typed as `undefined`

I'm not sure if this is a bug, or intended behavior. But it was surprising. If the type is just a tuple, indexing beyond its length produced `undefined`. But if it is part of an intersection, the element type is changed to `number | boolean`, as if it is no longer a tuple, but an array. However, indexing `0` or `1` still produces the correct type for those element positions.

### PR DESCRIPTION
**Pull request by jonhue**
Fixes #42557

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -21742,6 +21743,10 @@ namespace ts {
             return type.flags & TypeFlags.Union ? every((<UnionType>type).types, f) : f(type);
```
Comment (jonhue) [inline src/compiler/checker.ts:21743]: Is there a reason why this doesn't iterate over constituents of intersection types too?

File: tests/baselines/reference/checkJsxChildrenCanBeTupleType.errors.txt [hunk 2]
```diff
@@ -1,13 +1,8 @@
-tests/cases/conformance/jsx/checkJsxChildrenCanBeTupleType.tsx(17,18): error TS2769: No overload matches this call.
```
Comment (jonhue) [inline tests/baselines/reference/checkJsxChildrenCanBeTupleType.errors.txt:1]: I still need to fix this error.

Comment (jonhue) [reply tests/baselines/reference/checkJsxChildrenCanBeTupleType.errors.txt:1]: I did a lot of further testing and it seems like the reason why the previous errors disappear, is that in `checkApplicableSignatureForJsxOpeningLikeElement` we call `checkTypeRelatedTo` without an `errorNode` so this error is not raised.
If we pass the `errorNode` down, however, other errors change. I'm not sure what behavior we want here.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -14204,22 +14204,27 @@ namespace ts {
...
+                                error(indexNode, Diagnostics.Property_0_does_not_exist_on_type_1, unescapeLeadingUnderscores(propName), typeToString(objectType));
+                            }
                         }
+                        errorIfWritingToReadonlyIndex(getIndexInfoOfType(objectType, IndexKind.Number));
+                        return mapType(objectType, t => {
+                            const restType = getRestTypeOfTupleType(<TupleTypeReference>t) || undefinedType;
+                            return noUncheckedIndexedAccessCandidate ? getUnionType([restType, undefinedType]) : restType;
+                        });
+                    }
+                    else if (someType(objectType, isTupleType)) {
```
Comment (weswigham) [inline src/compiler/checker.ts:14225]: This returns `undefined` for something like `{[idx: number]: number} | [string]`, which is incorrect (the object type has a numeric indexer, after all) - at worst, you'd want something like `number | undefined` for an index of `1` or more, and not `undefined`. Moreover, an intersection, like `{[idx: number]: number} & [string]` should likewise produce `number` for the result of indexing `1`, and not `undefined`.

**Review by weswigham — CHANGES_REQUESTED** [review]

**Comment by sandersn** [comment]
Do you want to keep working on this? Otherwise I'd like to close it to reduce the number of open PRs.

**Comment by jonhue** [comment]
I'm sorry, I don't have the time at the moment.

**Pull request closed without merging**